### PR TITLE
Revert #1418, #1367, introduce public version of RestMethodInfo

### PR DESCRIPTION
**Pull request by anaisbetts**
Reverts #1418 and #1367, we will add the functionality of #1367 back as part of a fixed up version of #1317, and we'll re-evaluate #1418. This PR also removes access to the internal type RestMethodInfo, replacing it with a read-only sanitized version. While initially this public type will be limited, we can evaluate adding new fields as-needed. 

## TODO:

- [x] Revert #1418 
- [x] Revert #1367
- [x] Reimplement #1317 without using internal types
- [x] add functionality from #1367 in new public types as needed 
- [ ] Fix the merge commit that I have inevitably borked
- [ ] Docs / tests / etc

### CODE REVIEW

**Comment by anaisbetts** [comment]
@james-s-tayler Can you make sure that the change proposed here still works for your use-case?

**Comment by james-s-tayler** [comment]
@anaisbetts I think it will.

**Comment by cbrevik** [comment]
I was looking forward to the changes in #1418 being released, can I ask why it was reverted? Current name schemes coming from the `UniqueName`-implementation seemingly really bloats up logs / metrics.

**Comment by anaisbetts** [comment]
@cbrevik I reverted it because the implementation was too Java'y and it was extremely unclear from the PR what problem this actually solves or why it should exist - can you explain further?

**Comment by cbrevik** [comment]
The problem it solves relates to how other logging and metric libraries logs the name of the HTTP-client.

See for example #1467 in this repository and the issue it links back to from the Serilog-repository: https://github.com/serilog/serilog/issues/1793#issue-1521198225

Serilog pulls the `FullName` from the HttpClient which normally includes namespace + type, as per the comment: https://github.com/serilog/serilog/issues/1793#issuecomment-1372632571

Logging of "normal" HTTP-clients will be in this format:
`2023-01-05 18:45:53.898 +01:00 [INF] System.Net.Http.HttpClient.MyController.LogicalHandler Start processing HTTP request POST http://example.com/stuff`

Whilst logging of Refit-clients will be like this:
`2023-01-05 18:45:53.861 +01:00 [INF] System.Net.Http.HttpClient.Refit.Implementation.Generated+MyProjectIApi, MyProject, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null.LogicalHandler Start processing HTTP request POST http://example.com/stuff` 

Which including the prefixed stuff includes Version, PublicKeyToken etc.

Mostly then because of the implementation of https://github.com/reactiveui/refit/blob/main/Refit/UniqueName.cs

I've seen the same issue with for example using Prometheus, a sample metric from there:
`httpclient_response_duration_seconds_sum{method="GET",host="api.example.com",client="Refit.Implementation.Generated+MyFancyApiInfrastructureInsightClientIInsightHttpClient, My.Fancy.Api, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null",code="200"} 0.2517555`

So the issue is that it really bloats the logs and metrics up, and makes them a bit unreadable. Especially if you have (like we have) several different HTTP-clients with Refit. So I'd like for the generated name to be either shorter by default (and I see from your explanation why it shouldn't), or at least make it overridable if not (like in #1418 or a less Java-y solution)

**Comment by anaisbetts** [comment]
@cbrevik Got it, thanks for the explanation. I think we can figure something out then

**Comment by anaisbetts** [comment]
@cbrevik Can you confirm that https://github.com/reactiveui/refit/pull/1496/commits/3248979b46a3d800d40cdb7c1bfdfefac32431f8 works for you?

File: Refit/HttpRequestMessageProperties.cs [hunk 1]
```diff
@@ -12,24 +10,9 @@ public static class HttpRequestMessageOptions
...
-        /// <summary>
-        /// A typed key to access the <see cref="System.Type"/> of the top-level interface where the method was called from
-        /// on the <see cref="System.Net.Http.HttpRequestMessage.Options"/>.
-        /// </summary>
-        public static System.Net.Http.HttpRequestOptionsKey<System.Type> InterfaceTypeKey { get; } = new(InterfaceType);
-#endif
-
         /// <summary>
-        /// Returns the <see cref="Refit.RestMethodInfo"/> of the method that was called
+        /// Returns the <see cref="Refit.RestMethodInfoInternal"/> of the top-level interface
```
Comment (anaisbetts) [inline Refit/HttpRequestMessageProperties.cs:14]: Typo

**Comment by cbrevik** [comment]
> @cbrevik Can you confirm that [3248979](https://github.com/reactiveui/refit/commit/3248979b46a3d800d40cdb7c1bfdfefac32431f8) works for you?

It looks like it'll be able to do what I want yeah! I can test it too if a preview nuget-package is released too? Unless there is some other easy way for me to test out that commit specifically?

**Pull request merged**